src/authorization/access_control.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::shared_types::User;
+use crate::PublicKey;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::{collections::BTreeMap, hash::Hash};
+
+/// ===========================================================
+///  Access control of data type instances and their content.
+/// ===========================================================
+
+/// The type of access to the native data structures.
+#[derive(Clone, Copy, Debug, Hash, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub enum AccessType {
+    /// Read data, owners and permissions.
+    Read,
+    /// Append new values.
+    Append,
+    /// Insert new values.
+    Insert,
+    /// Soft-update existing values.
+    Update,
+    /// Soft-delete existing values.
+    Delete,
+    /// Hard-update existing values.
+    HardUpdate,
+    /// Hard-delete existing values.
+    HardDelete,
+    /// Modify permissions for other users.
+    ModifyPermissions,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub enum AccessList {
+    Public(PublicAccessList),
+    Private(PrivateAccessList),
+}
+
+impl From<PrivateAccessList> for AccessList {
+    fn from(list: PrivateAccessList) -> Self {
+        AccessList::Private(list)
+    }
+}
+
+impl From<PublicAccessList> for AccessList {
+    fn from(list: PublicAccessList) -> Self {
+        AccessList::Public(list)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub enum UserAccess {
+    Public(PublicUserAccess),
+    Private(PrivateUserAccess),
+}
+
+impl From<PrivateUserAccess> for UserAccess {
+    fn from(access: PrivateUserAccess) -> Self {
+        UserAccess::Private(access)
+    }
+}
+
+impl From<PublicUserAccess> for UserAccess {
+    fn from(access: PublicUserAccess) -> Self {
+        UserAccess::Public(access)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PrivateUserAccess {
+    status: BTreeMap<AccessType, bool>,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PublicUserAccess {
+    status: BTreeMap<AccessType, bool>,
+}
+
+impl PrivateUserAccess {
+    pub fn new(status: BTreeMap<AccessType, bool>) -> Self {
+        PrivateUserAccess { status }
+    }
+
+    pub fn set(&mut self, status: BTreeMap<AccessType, bool>) {
```

Comment:
I don't really see the point in having a `set` method when someone could just as easily create a new `UserAccess` instance. Also, making the `status` field public would seem to obviate the need for this as well as the `new` constructor.

Reply:
This is how it looked in Permissions already (different name though). But yeah, I agree, that's a bit strange actually, not so functional. 

The reason I think was that in the `PublicUserAccess`, there is logic for filtering out queries (originally, I did not implement that filter here yet).
But, that logic could just as well go in the ctor. 

> making the `status` field public would seem to obviate the need for this as well as the `new` constructor.
Public field wouldn't be desired here I'd say.

I think moving the query filter to ctor and always instantiate new instead of modifying an instance would be best here?

Reply:
Ah, that makes sense now. I didn't know there was going to be filtering logic.

> I think moving the query filter to ctor and always instantiate new instead of modifying an instance would be best here?

I agree!